### ISSUE TITLE
Wrong order of actions when removing user from group and deleting it in one go

### ISSUE THREAD

**Issue by 5030200** [fortinetdev/terraform-provider-fortios#204]
I have successfully created the following resources.

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a.name
  }
  member {
    name = fortios_user_local.b.name
  }
}

resource "fortios_user_local" "a" {
  name     = "a"
  status   = "enable"
  type     = "password"
}

resource "fortios_user_local" "b" {
  name     = "b"
  status   = "enable"
  type     = "password"
}
```

Now I remove user b:

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a.name
  }
}

resource "fortios_user_local" "a" {
  name     = "a"
  status   = "enable"
  type     = "password"
}
```

Running `terraform apply` fails because it tries to destroy the user before removing it from the group:

```
$ terraform apply
fortios_user_local.b: Refreshing state... [id=b]
fortios_user_local.a: Refreshing state... [id=a]
fortios_user_group.test: Refreshing state... [id=test]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place
  - destroy

Terraform will perform the following actions:

  # fortios_user_group.test will be updated in-place
  ~ resource "fortios_user_group" "test" {
        id                       = "test"
        name                     = "test"
        # (18 unchanged attributes hidden)

      - member {
          - name = "b" -> null
        }
        # (1 unchanged block hidden)
    }

  # fortios_user_local.b will be destroyed
  # (because fortios_user_local.b is not in configuration)
  - resource "fortios_user_local" "b" {
      - auth_concurrent_override  = "disable" -> null
      - auth_concurrent_value     = 0 -> null
      - authtimeout               = 0 -> null
      - fosid                     = 16777487 -> null
      - id                        = "b" -> null
      - name                      = "b" -> null
      - passwd_time               = "2022-01-09 12:00:00" -> null
      - sms_server                = "fortiguard" -> null
      - status                    = "enable" -> null
      - two_factor                = "disable" -> null
      - type                      = "password" -> null
      - username_case_sensitivity = "enable" -> null
    }

Plan: 0 to add, 1 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_user_local.b: Destroying... [id=b]
╷
│ Error: Error deleting UserLocal resource: Internal Server Error - Internal error when processing the request (500)
│ 
```

I am using fortinetdev/fortios v1.13.2. To work around the issue, I run `terraform apply -target=fortios_user_group.test && terraform apply` but it would be nice if the provider could handle the correct order.

Keep up the good work.

**Comment by poroping** [comment]
Try the lifecycle argument.

```hcl
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a.name
  }
  member {
    name = fortios_user_local.b.name
  }
}

resource "fortios_user_local" "a" {
  name     = "a"
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_user_local" "b" {
  name     = "b"
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}
```

**Comment by lix-fortinet** [comment]
Hi @poroping, 

Thank you for your response.

Hi @5030200,

Thank you for raising this issue. We could not handle the process sequence in the provider side. It is controlled by Terraform. The way we solve this issue is using lifecycle with argument create_before_destroy set to true, like @poroping suggested. However, based on our test, it only works on dynamic resources using meta-argument `count` or `for_each`. 

For your case, please try the following:

```
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a[0].name
  }
  member {
    name = fortios_user_local.b[0].name
  }
  depends_on = [
    fortios_user_local.a,
    fortios_user_local.b
  ]
}

resource "fortios_user_local" "a" {
  count = 1

  name     = "a"
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_user_local" "b" {
  count = 1

  name     = "b"
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}
```

When you want to delete resource `fortios_user_local.b`, you could set argument `count` of `fortios_user_local.b` to 0 and delete related member in `fortios_user_group`. For instance:
```
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a[0].name
  }

  depends_on = [
    fortios_user_local.a,
    fortios_user_local.b
  ]
}

resource "fortios_user_local" "a" {
  count = 1

  name     = "a"
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_user_local" "b" {
  count = 0

  name     = "b"
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}
```

If the only different between all users' config is the name, you could also use `locals` block and `for_each` keyword to implement it. And you just need to update `user_name_set` to add or remove users. For instance:

```
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  dynamic "member" {
    for_each = fortios_user_local.test
    content {
      name = member.value["name"]
    }
  }

  depends_on = [
    fortios_user_local.test
  ]
}

resource "fortios_user_local" "test" {
  for_each = local.user_name_set

  name     = each.value
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}

locals {
  user_name_set = toset(["a", "b"])
}
```

Please let me know if you have any questions.

Thanks,
Xing

**Comment by 5030200** [comment]
Many thanks for the suggestions. I am afraid I still have some trouble here.

When changing from

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a.name
  }
  member {
    name = fortios_user_local.b.name
  }
}

resource "fortios_user_local" "a" {
  name     = "a"
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_user_local" "b" {
  name     = "b"
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}
```

to

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  member {
    name = fortios_user_local.a.name
  }
  member {
    name = fortios_user_local.c.name
  }
}

resource "fortios_user_local" "a" {
  name     = "a"
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_user_local" "c" {
  name     = "b" # still b
  status   = "enable"
  type     = "password"

  lifecycle {
    create_before_destroy = true
  }
}
```

I get:

```
Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_user_local.b: Destroying... [id=b]
fortios_user_local.c: Creating...
Error: Error deleting UserLocal resource: Internal Server Error - Internal error when processing the request (500)
Error: Error creating UserLocal resource: Internal Server Error - Internal error when processing the request (500)
  on main.tf line 22, in resource "fortios_user_local" "c":
  22: resource "fortios_user_local" "c" {
```

When changing from

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  dynamic "member" {
    for_each = fortios_user_local.test
    content {
      name = member.value["name"]
    }
  }

  depends_on = [
    fortios_user_local.test
  ]
}

resource "fortios_user_local" "test" {
  for_each = local.user_name_set

  name     = each.value
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}

locals {
  user_name_set = toset(["a", "b"])
}
```

to

```tf
resource "fortios_user_group" "test" {
  name       = "test"
  group_type = "firewall"
  dynamic "member" {
    for_each = fortios_user_local.test
    content {
      name = member.value["name"]
    }
  }

  depends_on = [
    fortios_user_local.test
  ]
}

resource "fortios_user_local" "test" {
  for_each = local.user_name_set

  name     = each.value
  status   = "enable"
  type     = "password"
  lifecycle {
    create_before_destroy = true
  }
}

locals {
  user_name_set = toset(["c", "d"])
}
```

I get:

```
Plan: 2 to add, 1 to change, 2 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_user_local.test["d"]: Creating...
fortios_user_local.test["c"]: Creating...
fortios_user_local.test["d"]: Creation complete after 0s [id=d]
fortios_user_local.test["c"]: Creation complete after 0s [id=c]

Error: Provider produced inconsistent final plan

When expanding the plan for fortios_user_group.test to include new values
learned so far during apply, provider
"registry.terraform.io/fortinetdev/fortios" produced an invalid new value for
.member: block count changed from 2 to 4.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

I actually already use a list together with for_each but I tried to present the issue as simple as possible.

**Comment by lix-fortinet** [comment]
Hi @5030200,

As for your first scenario, please try to use meta-argument  `count`. Remember the index `[0]` in the member block and block `depends_on` is also required. 

As for your second scenario, it works in my end:
```
$ terraform apply --auto-approve
fortios_user_local.test["b"]: Refreshing state... [id=b]
fortios_user_local.test["a"]: Refreshing state... [id=a]
fortios_user_group.test: Refreshing state... [id=test]
fortios_user_local.test["c"]: Creating...
fortios_user_local.test["d"]: Creating...
fortios_user_local.test["d"]: Creation complete after 0s [id=d]
fortios_user_local.test["c"]: Creation complete after 0s [id=c]
fortios_user_group.test: Modifying... [id=test]
fortios_user_group.test: Modifications complete after 0s [id=test]
fortios_user_local.test["b"]: Destroying... [id=b]
fortios_user_local.test["a"]: Destroying... [id=a]
fortios_user_local.test["a"]: Destruction complete after 0s
fortios_user_local.test["b"]: Destruction complete after 0s

Apply complete! Resources: 2 added, 1 changed, 2 destroyed.
```

Could I know your Terraform version? `lifecycle` block requires Terraform v1.14 or later.

Thanks,
Xing

**Comment by 5030200** [comment]
Thanks a lot for your help! I restructured things a bit and got it working. `lifecycle` block seems available since Terraform v0.12 or maybe even earlier.
```tf
locals {
  groups = {
    "test" = [
      "a",
      "b",
    ]
  }
  users = [
    "a",
    "b",
  ]
}

resource "fortios_user_group" "group" {
  for_each = local.groups
  name     = each.key
  dynamic "member" {
    for_each = each.value
    content {
      name = fortios_user_local.user[member.value].id
    }
  }
}

resource "fortios_user_local" "user" {
  lifecycle {
    create_before_destroy = true
  }
  for_each = local.users
  name     = each.key
  status   = "enable"
  type     = "password"
}
```
